### PR TITLE
Move best practices out of "Type Stubs" doc into guide on writing stubs.

### PR DESCRIPTION
**Pull request by rchen152**
This is part of an effort to integrate the standalone "Type Stubs" doc into the rest of the typing documentation. See
https://github.com/python/typing/pull/1815 for more context.

For https://github.com/python/typing/issues/1605.

### CODE REVIEW

File: docs/guides/writing_stubs.rst [hunk 1]
```diff
@@ -106,3 +106,487 @@ particularly with false positives.
...
+  @type_check_only
+  class Readable(Protocol):
+      def read(self) -> str: ...
+
+  def get_reader() -> Readable: ...
+
+Structural Types
+----------------
+
+As seen in the example with ``_Readable`` in the previous section, a common use
```
Comment (JelleZijlstra) [inline docs/guides/writing_stubs.rst:193]:
```suggestion
As seen in the example with ``Readable`` in the previous section, a common use
```

File: docs/guides/writing_stubs.rst [hunk 2]
```diff
@@ -106,3 +106,487 @@ particularly with false positives.
...
+          print(y)
+
+  def maybe_raise(x):
+      """Raise an error if x (a boolean) is true."""
+      if x:
+          raise ValueError()
+
+The implementation of ``print_elements`` takes any iterable, despite the
+documented type of ``list``. In this case, annotate the argument as
+``Iterable[Any]``, to follow the :ref:`best practice<argument-return-practices>`
```
Comment (JelleZijlstra) [inline docs/guides/writing_stubs.rst:349]:
```suggestion
``Iterable[object]``, to follow the :ref:`best practice<argument-return-practices>`
```

File: docs/guides/writing_stubs.rst [hunk 3]
```diff
@@ -106,3 +106,487 @@ particularly with false positives.
...
+
+    def to_int1(x: str) -> int:
+        return int(x)
+    def to_int2(x: str) -> int:
+        ...
+    def to_int3(x: str) -> int: pass
+
+.. _private-definitions:
+
+Private Definitions
```
Comment (JelleZijlstra) [inline docs/guides/writing_stubs.rst:530]: Duplicates a previous section

Comment (rchen152) [reply docs/guides/writing_stubs.rst]: I consolidated this with the "Stub-Only Objects" section.

File: docs/guides/writing_stubs.rst [hunk 4]
```diff
@@ -106,3 +106,487 @@ particularly with false positives.
...
+
+Structural Types
+----------------
+
+As seen in the example with ``_Readable`` in the previous section, a common use
+of stub-only objects is to model types that are best described by their
+structure. These objects are called protocols (:pep:`544`), and it is encouraged
+to use them freely to describe simple structural types.
+
+It is `recommended <#private-definitions>`_ to prefix stub-only object names with ``_``.
```
[the remark this answers is not in the document]
Comment (rchen152) [reply docs/guides/writing_stubs.rst]: I deleted this sentence and reworded the "Stub-Only Objects" section to say that names should be underscore-prefixed unless you want to expose them to users, in which case use `@type_check_only`.

File: docs/guides/writing_stubs.rst [hunk 5]
```diff
@@ -106,3 +106,487 @@ particularly with false positives.
...
+----------------
+
+Partial stubs can be useful, especially for larger packages, but they should
+follow the following guidelines:
+
+* Included functions and methods should list all arguments, but the arguments
+  can be left unannotated.
+* Do not use ``Any`` to mark unannotated or partially annotated values. Leave
+  function parameters and return values unannotated and mark the function with
+  an ``# incomplete`` comment. This comment is mainly intended as a reminder for
```
[the remark this answers is not in the document]
Comment (rchen152) [reply docs/guides/writing_stubs.rst]: Delorted.

File: docs/guides/writing_stubs.rst [hunk 6]
```diff
@@ -106,3 +106,478 @@ particularly with false positives.
...
+Stub Content
+============
+
+This section documents best practices on what elements to include or
+leave out of stub files.
+
+Modules excluded fom stubs
+--------------------------
+
+Not all modules should be included into stubs.
```
Comment (srittau) [inline docs/guides/writing_stubs.rst:119]: Not a native speaker, but "into stubs" sounds wrong to be. "in stubs"?

**Review by srittau — APPROVED** [review]

File: docs/guides/writing_stubs.rst [hunk 7]
```diff
@@ -106,3 +106,478 @@ particularly with false positives.
...
+because the stub is wrong. Although it may also be helpful for a type
+checker to point out usage of private objects, false negatives (no errors for
+wrong code) are preferable over false positives (type errors
+for correct code). In addition, even for private objects a type checker
+can be helpful in pointing out that an incorrect type was used.
+
+``__all__``
+------------
+
+A stub file should contain an ``__all__`` variable if and only if it also
```
Comment (JelleZijlstra) [inline docs/guides/writing_stubs.rst:167]:
```suggestion
A stub file should contain an ``__all__`` variable if and only if it is also
```

File: docs/guides/writing_stubs.rst [hunk 8]
```diff
@@ -106,3 +106,478 @@ particularly with false positives.
...
+present at runtime. In that case, the contents of ``__all__`` should be
+identical in the stub and at runtime. If the runtime dynamically adds
+or removes elements (for example if certain functions are only available on
+some platforms), include all possible elements in the stubs.
```
Comment (JelleZijlstra) [inline docs/guides/writing_stubs.rst:171]: ... or use sys.platform checks to replicate the runtime logic. Let's use an example that can't be represented in stubs:
```suggestion
some system configurations), include all possible elements in the stubs.
```

File: docs/guides/writing_stubs.rst (hunk 6 above)
Comment (rchen152) [reply docs/guides/writing_stubs.rst]: Fixed, thanks!

**Comment by rchen152** [comment]
@srittau @JelleZijlstra Thanks for the reviews! Is there anyone else you'd recommend I request a review from or any other procedural steps to take before merging?

**Review by JelleZijlstra — APPROVED** [review]

**Comment by srittau** [comment]
I think this is fine as is. This is not a spec,  but an Informal document, so we can always easily amend it if something comes up.

**Review by hauntsaninja — APPROVED** [review]
(some of the content feels a little typeshed tailored, but improving that is separate from the move)

**Pull request merged**